src/engine/client/keys.h

```diff
@@ -32,40 +32,141 @@ Maryland 20850 USA.
 ===========================================================================
 */
 
+#ifndef COMMON_KEYS_H_
+#define COMMON_KEYS_H_
+
+#include <unordered_map>
+
 #include "keycodes.h"
 
 #include "framework/ConsoleField.h"
+#include "qcommon/q_unicode.h"
 
 #define MAX_TEAMS 4
 #define DEFAULT_BINDING 0
 
 struct qkey_t
 {
-	bool down;
-	int      repeats; // if > 1, it is autorepeating
-	char     *binding[ MAX_TEAMS ];
+    bool down;
+    int      repeats; // if > 1, it is autorepeating
+    char     *binding[ MAX_TEAMS ];
+};
+
+namespace Keyboard {
+class Key
+{
+public:
+    enum class Kind {
+        INVALID,      // No key.
+        UNICODE_CHAR, // The key that types this char (layout-dependent)
+        KEYNUM,       // keyNum_t values, for virtual keys that aren't characters
+                      //    Also used for mouse buttons
+        SCANCODE,     // "Physical" keys, these are encouraged over UNICODE_CHAR because SDL is
+                      //     inconsistent about whether it knows the Unicode character that a key
+                      //     maps to, or just defaults to the QWERTY key at that location
+        CONSOLE,      // Hack to integrate with existing code
+    };
+private:
+    Kind kind_;
+    int id_;
+    constexpr Key(Kind kind, int id): kind_(kind), id_(id) { }
+
+public:
+    static const Key NONE;
+    static const Key CONSOLE;
+
+    Key(): Key(Kind::INVALID, 0) { }
+    Key(keyNum_t key) {
```

Comment:
explicit Key(keyNum_t key)

I know it's syntactic sugar to convert a keyNum_t into a Key and you use it heavily for SDL, but it makes the code unclear. You can create the Key in QueueEvent and still keep the same syntactic sugar.